apps/ejabberd/src/mongoose_api_common.erl

```diff
@@ -0,0 +1,325 @@
+%%%-------------------------------------------------------------------
+%%% @author ludwikbukowski
+%%% @copyright (C) 2016, <COMPANY>
+%%% @doc
+%%%
+%%% @end
+%%% Created : 20. Jul 2016 10:16
+%%%-------------------------------------------------------------------
+-module(mongoose_api_common).
+-author("ludwikbukowski").
+-include("mongoose_api.hrl").
+-include("ejabberd.hrl").
+%%
+%% @doc MongooseIM REST API backend
+%% This module handles the client HTTP REST requests, then respectively convert them to Commands from mongoose_commands
+%% and execute with `admin` privileges.
+%% It supports responses with appropriate HTTP Status codes returned to the client.
+%% This module implements behaviour introduced in ejabberd_cowboy which is built on top of the cowboy library.
+%% The method supported: GET, POST, PUT, DELETE. Only JSON format.
+%% The library "jiffy" used to serialize and deserialized JSON data.
+%%
+%% %% Handling the requests:
+%% TODO
+
+
+%% API
+-export([create_admin_url_path/1,
+         create_user_url_path/1,
+         error_response/3,
+         error_response/4,
+         action_to_method/1,
+         method_to_action/1,
+         get_allowed_methods/1,
+         process_request/4,
+         reload_dispatches/1]).
+
+
+%% @doc Reload all ejabberd_cowboy listeners.
+%% When a command is registered or unregistered, the routing paths that
+%% cowboy stores as a "dispatch" must be refreshed.
+%% Read more http://ninenines.eu/docs/en/cowboy/1.0/guide/routing/
+reload_dispatches(_Command) ->
+    Listeners = supervisor:which_children(ejabberd_listeners),
+    CowboyListeners = [Child || {_Id, Child, _Type, [ejabberd_cowboy]}  <- Listeners],
+    [ejabberd_cowboy:reload_dispatch(Child) || Child <- CowboyListeners].
+
+
+-spec create_admin_url_path(mongoose_command()) -> ejabberd_cowboy:path().
+create_admin_url_path(Command) ->
+    "/" ++ category_to_resource(?COMMANDS_ENGINE:category(Command))
+           ++ maybe_add_bindings(Command, admin).
+
+-spec create_user_url_path(mongoose_command()) -> ejabberd_cowboy:path().
+create_user_url_path(Command) ->
+    "/" ++ category_to_resource(?COMMANDS_ENGINE:category(Command))
+           ++ maybe_add_bindings(Command, user).
+
+-spec process_request(method(), mongoose_command(), any(), #http_api_state{}) -> {any(), any(), #http_api_state{}}.
+process_request(<<"POST">>, Command, Req, #http_api_state{bindings = Binds, entity = Entity} = State) ->
+    BindsReversed = lists:reverse(Binds),
+    case parse_request_body(Req) of
+        {error, _R}->
+            error_response(bad_request, ?BODY_MALFORMED , Req, State);
+        {Params, Req2} ->
+            Params2 = BindsReversed ++ Params ++ maybe_add_caller(Entity),
+            handle_request(Command, Params2, Req2, State)
+    end;
+process_request(<<"PUT">>, Command, Req, #http_api_state{bindings = Binds, entity = Entity} = State) ->
```

Comment:
since the branch for `put` method looks actually quite the same as for `post` (we decided that we allow "identifiers" also for `post` methods), those two branches can be merged